### ISSUE TITLE
Vertically extended disk not working

### ISSUE THREAD

**Issue by kslong** [sirocco-rt/sirocco#160]
We have a mode intended to allow a vertically extended list, mode 2 of disk_type.  We are apparently unable to construct photons in this mode.  The program seems simply to get stuck in some kind of infinite loop.  There is the .pf file which I used.

Wind_type(0=SV,1=Sphere,2=Previous,3=Hydro,4=Corona,5=knigge,6=homologous,7=yso,8=elvis,9=shell)                    0
Atomic_data                         data/standard78
photons_per_cycle                              1e6
Ionization_cycles                               10
spectrum_cycles                                   1
Coord.system(0=spherical,1=cylindrical,2=spherical_polar,3=cyl_var)                    1
Wind.dim.in.x_or_r.direction                     30
Wind.dim.in.z_or_theta.direction                   30
Wind_ionization(0=on.the.spot,1=LTE,2=fixed,3=recalc_bb,6=pairwise_bb,7=pairwise_pow,8=matrix_bb,9=matrix_pow)                    3
Line_transfer(0=pure.abs,1=pure.scat,2=sing.scat,3=escape.prob,6=macro_atoms,7=macro_atoms+aniso.scattering)                    3
Thermal_balance_options(0=everything.on,1=no.adiabatic)                    0
System_type(0=star,1=binary,2=agn)                    0
disk.type(0=no.disk,1=standard.flat.disk,2=vertically.extended.disk)                    2
Star_radiation(y=1)                               1
Disk_radiation(y=1)                               1
Boundary_layer_radiation(y=1)                     0
Wind_radiation(y=1)                              1
Rad_type_for_star(0=bb,1=models)_to_make_wind                    0
Rad_type_for_disk(0=bb,1=models)_to_make_wind                    0
mstar(msol)                                     0.8
rstar(cm)                                     7e+08
tstar                                         40000
disk.mdot(msol/yr)                            1e-08
Disk.illumination.treatment(0=no.rerad,1=high.albedo,2=thermalized.rerad,3=analytic)                    0
Disk.temperature.profile(0=standard;1=readin)                    0
disk.radmax(cm)                             2.4e+10
wind.radmax(cm)                               1e+11
wind.t.init                                   40000
wind.mdot(msol/yr)                            1e-09
sv.diskmin(units_of_rstar)                        4
sv.diskmax(units_of_rstar)                       12
sv.thetamin(deg)                                 20
sv.thetamax(deg)                                 65
sv.mdot_r_exponent                                0
sv.v_infinity(in_units_of_vescape                    3
sv.acceleration_length(cm)                    7e+10
sv.acceleration_exponent                        1.5
wind.filling_factor(1=smooth,<1=clumped)                    1
Rad_type_for_star(0=bb,1=models,2=uniform)_in_final_spectrum                    0
Rad_type_for_disk(0=bb,1=models,2=uniform)_in_final_spectrum                    0
spectrum_wavemin                              850
spectrum_wavemax                              1850
no_observers                                      4
angle(0=pole)                                    10
angle(0=pole)                                    30
angle(0=pole)                                    60
angle(0=pole)                                    80
live.or.die(0).or.extract(anything_else)                    1
spec.type(flambda(1),fnu(2),basic(other)                    1
reverb.type                                       0
Photon.sampling.approach(0=T,1=(f1,f2),2=cv,3=yso,4=user_defined,5=cloudy_test,6=wide,7=AGN,8=logarithmic)                    2

**Comment by jhmatthews** [comment]
@kslong what's the error message and what are you using for z0 and power law index?

**Comment by kslong** [comment]
I get a segmentation fault.

Wind_type(0=SV,1=Sphere,2=Previous,3=Hydro,4=Corona,5=knigge,6=homologous,7=yso,8=elvis,9=shell)                    0
Atomic_data                         data/standard78
photons_per_cycle                              1e6
Ionization_cycles                               10
spectrum_cycles                                   1
Coord.system(0=spherical,1=cylindrical,2=spherical_polar,3=cyl_var)                    3
Wind.dim.in.x_or_r.direction                     30
Wind.dim.in.z_or_theta.direction                   30
Wind_ionization(0=on.the.spot,1=LTE,2=fixed,3=recalc_bb,6=pairwise_bb,7=pairwise_pow,8=matrix_bb,9=matrix_pow)                    3
Line_transfer(0=pure.abs,1=pure.scat,2=sing.scat,3=escape.prob,6=macro_atoms,7=macro_atoms+aniso.scattering)                    3
Thermal_balance_options(0=everything.on,1=no.adiabatic)                    0
System_type(0=star,1=binary,2=agn)                    0
disk.type(0=no.disk,1=standard.flat.disk,2=vertically.extended.disk)                    1
Star_radiation(y=1)                               1
Disk_radiation(y=1)                               1
Boundary_layer_radiation(y=1)                     0
Wind_radiation(y=1)                              1
Rad_type_for_star(0=bb,1=models)_to_make_wind                    0
Rad_type_for_disk(0=bb,1=models)_to_make_wind                    0
mstar(msol)                                     0.8
rstar(cm)                                     7e+08
tstar                                         40000
disk.mdot(msol/yr)                            1e-08
Disk.illumination.treatment(0=no.rerad,1=high.albedo,2=thermalized.rerad,3=analytic)                    0
Disk.temperature.profile(0=standard;1=readin)                    0
disk.radmax(cm)                             2.4e+10
wind.radmax(cm)                               1e+11
wind.t.init                                   40000
wind.mdot(msol/yr)                            1e-09
sv.diskmin(units_of_rstar)                        4
sv.diskmax(units_of_rstar)                       12
sv.thetamin(deg)                                 20
sv.thetamax(deg)                                 65
sv.mdot_r_exponent                                0
sv.v_infinity(in_units_of_vescape                    3
sv.acceleration_length(cm)                    7e+10
sv.acceleration_exponent                        1.5
wind.filling_factor(1=smooth,<1=clumped)                    1
Rad_type_for_star(0=bb,1=models,2=uniform)_in_final_spectrum                    0
Rad_type_for_disk(0=bb,1=models,2=uniform)_in_final_spectrum                    0
spectrum_wavemin                              850
spectrum_wavemax                              1850
no_observers                                      4
angle(0=pole)                                    10
angle(0=pole)                                    30
angle(0=pole)                                    60
angle(0=pole)                                    80
live.or.die(0).or.extract(anything_else)                    1
spec.type(flambda(1),fnu(2),basic(other)                    1
reverb.type                                       0
Photon.sampling.approach(0=T,1=(f1,f2),2=cv,3=yso,4=user_defined,5=cloudy_test,6=wide,7=AGN,8=logarithmic)                    2
~

The weird thing about this is that I seem to have put a flat disk in, so maybe this involves logically inconsistent inputs (that the program does not catch).

**Comment by jhmatthews** [comment]
Hunch: I bet this is due to moving the order of the parameter file around (disk.type got moved to after system type, before it was specified later).

**Comment by kslong** [comment]
We need to recheck that this is really a problem, as I did do some work on this, and as far as I can remember I thought the vertically extended disk is working.   We need some test cases for this.

**Comment by kslong** [comment]
Re-tested this today.  My first tests made it look as this was working OK.  However, ultimately I was able to reproduce the original problem (Not the SegFault) with this .pf file, running in single processor mode and with gcc and debug on my Mac.
[test.pf.txt](https://github.com/agnwinds/python/files/1765100/test.pf.txt)

The error is quite weird, in the sense that it is occurring nearly line 3221 of get_atomic_data, where because f is nan one gets trapped.  

There is an earlier error dot:sane check which is probably which created the nan.  This is ultimately coming from photo_gen_disk (which provides a connection as to why this might be occurring for this disk)  The offending line is proabably line 1196 of photo_gen_disk, where theta is a NAN, even though the values that go into zdisk look plausible.

We need to understand the original error, and probably fix loop in get_atomic_data to prevent the infinite loop.

**Comment by kslong** [comment]
Having solved the proximate cause of the hang, I looked at whether the routines associated with a vertically extended disk were actually working.  I found serious problems.  First, for a vertically extended disk, we are not identifying photons as hitting the disk, which originate in the bowl of the disk.  Our test for whether a photon hits the disk were to set up a plane above and below the disk and see whether photons intersected one or the other of these planes inside the disk radius.  Unless this condition was satisfied, we did not actually check whether the photon encounter the disk surface.  

Furthermore, ds_to_disk is problematic, because in some cases, e.g. for determining how far it is from a position in the wind to the footpoint of the wind, we allow "negative" distances, but for the calculation of photons traversing the wind, we only want positive values.  There is even an ambiguous situation where you can have a photon located down in the bowl with a flat trajectory, that can hit it both in the positive and negative direction.

**Comment by kslong** [comment]
I have now updated walls and ds_to_disk so that the distance to the disk is properly calculated for a vertically extended disk.  In the process, I also clarified the logic for when ds_to_disk should be allowed to return a negative value.  The parameter file test.pf now produces a plots for total number of photons in the wind, that look likes like the following:

![test_ntot](https://user-images.githubusercontent.com/2530901/36877211-ad4c18ae-1d7e-11e8-990e-685a3cf2042b.png)
The plot above only includes cells that are fully in the wind:

Another plot for a model with 100 x 100 cells, where I have included cells with are paritally int he wind as well, and have used a larger star,   looks like the following:

![test3_log_ntot](https://user-images.githubusercontent.com/2530901/36877272-f51ef4c6-1d7e-11e8-9acc-ce26edb2d3d8.png)

It's worthile noticing in the case, that we do not have the ideal gridding near the base of the wind.  this is a problem that oen might want to investigate but it's not directly related to this issue, which I am now closing.  

I have also merged these changes back into dev.
